src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,183 @@
 module Juvix.Backends.ArithmeticCircuit.Compilation.Types where
 
-import qualified Data.Graph.Inductive as G
-import Juvix.Library
-
-type Circuit = G.Gr NodeType WireType
-
-data NodeType
-  = -- Addition gate.
-    Add
-  | -- Multiplication gate.
-    Mul
-  | -- Input value (by vector index).
-    InputValue Natural
-  | -- Output value (by vector index).
-    OutputValue Natural
-  deriving (Show, Eq, Generic)
-
-type WireType = ()
+import qualified Juvix.Core.ErasedAnn as CoreErased
+import qualified Juvix.Core.Usage as Usage
+import qualified Juvix.Library as J
+import qualified Juvix.Backends.ArithmeticCircuit.Parameterisation as Par
+import qualified Circuit.Expr as Expr
+import qualified Circuit.Lang as Lang
+import qualified Circuit
+import qualified Data.Map as Map
+import Numeric.Natural()
+
+import Protolude
+
+data PrimVal
+  = Element Par.F
+  | Boolean Bool
+  | FEInteger Int
+  | BinOp BinOp Term Term
+  | Op Op Term
+  | If Term Term Term
+  deriving (Show, Eq)
+
+data BinOp
+  = Add
+  | Mul
+  | Sub
+  | Exp
+  | Eq
+  | And
+  | Or
+  deriving (Show, Eq)
+
+data Op = Neg
+  deriving (Show, Eq)
+
+type Term = CoreErased.AnnTerm () PrimVal
```

Comment:
This needs to be `CoreErased.AnnTerm PrimTy PrimVal`, for some `PrimTy`, which needs to be defined

We should be able to construct a `Core.Parameterisation` record (example - https://github.com/metastatedev/juvix/blob/develop/src/Juvix/Backends/Michelson/Parameterisation.hs#L139)

It's fine not to worry about the parsing parts for now, but we will/do need type-checking

Reply:
... you have a `Ty` defined in the other file which seems correct

Reply:
... so basically this needs to be the same type as in `Parameterisation.hs`, since that's what we compile from.

It's fine to use other types "inside" the compilation logic, if convenient

Reply:
I haven't do any typechecking for it yet. Glad to add it now 🙂